### PR TITLE
[mlir] Replace dynamic sizes in insert_slice of tensor.cast canonicalization

### PR DESCRIPTION
**Pull request by Max191**
In some cases this pattern may ignore static information due to dynamic operands in the insert_slice sizes operands, e.g.:
```
%0 = tensor.cast %arg0 : tensor<1x?xf32> to tensor<?x?xf32>
%1 = tensor.insert_slice %0 into %arg1[...] [%s0, %s1] [...] 
    : tensor<?x?xf32> into tensor<?x?xf32>
```
Can be rewritten into:
```
%1 = tensor.insert_slice %arg0 into %arg1[...] [1, %s1] [...] 
    : tensor<1x?xf32> into tensor<?x?xf32>
```
This PR updates the matching in the pattern to allow rewrites like this.

### CODE REVIEW

File: mlir/lib/Dialect/Tensor/IR/TensorOps.cpp [hunk 1]
```diff
@@ -2711,15 +2711,38 @@ struct InsertSliceOpCastFolder final : public OpRewritePattern<InsertOpTy> {
     auto dstType = llvm::dyn_cast<RankedTensorType>(dst.getType());
     if (!srcType || !dstType)
       return failure();
+
+    // The tensor.cast source could have additional static information not seen
+    // in the insert slice op static sizes, so we ignore dynamic dims when
+    // computing the rank reduction mask.
+    SmallVector<int64_t> staticSizes(insertSliceOp.getStaticSizes());
```
Comment (hanhanW) [inline mlir/lib/Dialect/Tensor/IR/TensorOps.cpp:2718]: The values won't be modified, how about using `ArrayRef<int64_t>` here?

**Review by hanhanW — APPROVED** [review]
thanks for the improvement!

File: mlir/lib/Dialect/Tensor/IR/TensorOps.cpp (hunk 1 above)
Comment (Max191) [reply mlir/lib/Dialect/Tensor/IR/TensorOps.cpp:2718]: They actually do get modified later on L2735.

Comment (hanhanW) [reply mlir/lib/Dialect/Tensor/IR/TensorOps.cpp:2718]: I see, somehow I missed it when I was searching the use of the variable.

File: mlir/test/Dialect/Tensor/canonicalize.mlir [hunk 2]
```diff
@@ -1890,21 +1918,6 @@ func.func @splat_dynamic_no_fold(%m: index) -> tensor<4x?xf32> {
 
 // -----
 
-// There was an issue in cast + insert_slice folding generating invalid ir.
-// https://github.com/llvm/llvm-project/issues/53099
```
Comment (nicolasvasilache) [inline mlir/test/Dialect/Tensor/canonicalize.mlir:1894]: hmm this test was guarding against an issue that was previously fixed. 
Why is it removed?

Comment (hanhanW) [reply mlir/test/Dialect/Tensor/canonicalize.mlir:1894]: It is moved to l.758. I think the revision adds the support for the case. It is generating valid IR now.

Comment (nicolasvasilache) [reply mlir/test/Dialect/Tensor/canonicalize.mlir:1894]: ah yes, thank you!

**Pull request merged**